src/libexpr/include/nix/expr/value.hh

```diff
@@ -643,6 +748,65 @@ protected:
     {
         setUntaggablePayload<pdPath>(path.accessor, path.path);
     }
+
+    void setStorage(Failed * failed) noexcept
+    {
+        setSingleDWordPayload<tFailed>(std::bit_cast<PackedPointer>(failed));
+    }
+
+    ValueStorage() {}
+
+    ValueStorage(const ValueStorage & v)
+    {
+        *this = v;
+    }
+
+    /**
+     * Copy a value. This is not allowed to be a thunk to avoid
+     * accidental work duplication.
+     */
+    ValueStorage & operator=(const ValueStorage & v)
+    {
+        auto p1_ = v.p1;
+        auto p0_ = v.p0.load(std::memory_order_acquire);
```

Comment:
Do we also want to include the comment of "p1 must be written to before p0" here?